### PR TITLE
feat(test): add etcd simulator

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR adds an etcd simulator to deterministic e2e test, and switches the meta backend from 'mem' to 'etcd'.

The [implementation](https://github.com/madsim-rs/madsim/blob/e40c7caf4a94410e22f856d6a754cc5d99d51b37/madsim-etcd-client/src/service.rs) of this simulator is very similar to the `MemStore` in risingwave. But it's running behind the 'network' and follows the etcd API. This will help us check whether we are correctly using these API, and more importantly, handling the unexpected behavior from network and etcd server.

For instance, it can simulate the "etcdserver: request timed out" as we encountered earlier at #4919. This feature is not enabled by default since we have not fixed the bug yet. If you turn this on you will find the meta node panics somewhere. Hope it would be helpful with debugging.

This PR also updates madsim, which fixes memory leak in recovery test and forbids creating system thread.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#4919

### CODE REVIEW

**Review by skyzh — APPROVED** [review]
🤤🤤🤤

File: src/tests/simulation/src/main.rs [hunk 1]
```diff
@@ -109,6 +109,26 @@ async fn main() {
...
+        .create_node()
+        .name("etcd")
+        .ip("192.168.10.1".parse().unwrap())
+        .init(|| async {
+            let addr = "0.0.0.0:2388".parse().unwrap();
+            etcd_client::SimServer::builder()
+                // FIXME: If etcd returns timeout error, the meta will panic.
+                // To debug, change this rate to 0.1 and the sleep duration after meta to 30s.
+                // Tracking issue: https://github.com/risingwavelabs/risingwave/issues/4919
+                .timeout_rate(0.0)
```
Comment (BugenZhao) [inline src/tests/simulation/src/main.rs:123]: FYI, we've added the retry logic for etcd client in #5056. It should be okay to set a non-zero rate now. :)

**Review by BugenZhao — APPROVED** [review]

**Comment by wangrunji0408** [comment]
Nexmark_q7 failed after etcd timeout enabled:
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: test error at e2e_test/streaming/nexmark_endless.slt:10: query result mismatch:
--
  | [SQL] select count(*) > 0 as has_results from nexmark_q7;
  | [Diff]
  | t
  | f
```
Is this a known issue? cc @BugenZhao

**Comment by BugenZhao** [comment]
> Nexmark_q7 failed after etcd timeout enabled: https://buildkite.com/risingwavelabs/pull-request/builds/7618#01830c5a-65e3-4caf-8398-866600185b9a/812-839
> 
> ```
> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: test error at e2e_test/streaming/nexmark_endless.slt:10: query result mismatch:
> --
>   | [SQL] select count(*) > 0 as has_results from nexmark_q7;
>   | [Diff]
>   | t
>   | f
> ```
> 
> Is this a known issue? cc @BugenZhao

This case tests the generation of nexmark source and it's unstable these days. cc @fuyufjh 

However, does the system trigger recovery? If so, It may also be caused by source not correctly recovered.

**Comment by wangrunji0408** [comment]
> However, does the system trigger recovery?

No. The only thing is the etcd request may block for 5~15s and then get timeout with a 1% probability.

**Comment by wangrunji0408** [comment]
BAD NEWS: I've found several problems for the following test on my mac:

```sh
./risedev sslt --release -- "e2e_test/streaming/**/*.slt"
```

MADSIM_TEST_SEED=2,3,4,16,19
```
thread '<unnamed>' panicked at 'Failed to connect to database: Error { kind: Connect, cause: Some(Custom { kind: ConnectionRefused, error: "connection refused" }) }', src/tests/simulation/src/main.rs:365:14
```

MADSIM_TEST_SEED=9
```
thread '<unnamed>' panicked at 'no entry found for key', src/meta/src/stream/scheduler.rs:98:23
```

MADSIM_TEST_SEED=1,15
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: test error at e2e_test/streaming/datagen.slt:18: query result mismatch:
[SQL] select v1, v2 from s1 where v1 is not null limit 15;
[Diff]
1 11
10 20
2 12
3 13
4 14
5 15
6 16
7 17
8 18
9 19
```

MADSIM_TEST_SEED=17
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: test error at e2e_test/streaming/datagen.slt:48: query result mismatch:
[SQL] select v1 from s1 order by v1 limit 10;
[Diff]
1
2
3
4
5
6
7
8
9
10
```

MADSIM_TEST_SEED=6,8,10,11,12,13,18
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: test error at e2e_test/streaming/nexmark_endless.slt:10: query result mismatch:
[SQL] select count(*) > 0 as has_results from nexmark_q7;
[Diff]
t
f
```

MADSIM_TEST_SEED=5,7,14,20
```
Pass
```

**Pull request merged**